pkg/controller/nodecsrapprover/node_csr_approver.go

```diff
@@ -130,40 +153,104 @@ func (r *reconciler) reconcile(ctx context.Context, request reconcile.Request) e
+		return "", fmt.Errorf("node name is empty")
 	}
-	return false
-}
 
-// validateCSRRequest parses the CSR request and compares certificate request CommonName with CSR username, and
-// certificateRequest Organization with CSR groups.
-func validateCSRRequest(csr *certificatesv1beta1.CertificateSigningRequest) error {
-	csrBlock, rest := pem.Decode(csr.Spec.Request)
-	if csrBlock == nil {
-		return fmt.Errorf("no certificate request found for the given CSR")
+	// Ensure system:nodes and system:authenticated are in groups
+	if len(csr.Spec.Groups) < 2 {
+		return "", fmt.Errorf("there are less than 2 groups")
 	}
-	if len(rest) != 0 {
-		return fmt.Errorf("found more than one PEM encoded block in the result")
+	if !sets.NewString(csr.Spec.Groups...).HasAll(nodeGroup, authenticatedGroup) {
+		return "", fmt.Errorf("'%s' and/or '%s' are not in its groups", nodeGroup, authenticatedGroup)
 	}
-	csrRequest, err := x509.ParseCertificateRequest(csrBlock.Bytes)
-	if err != nil {
-		return err
+
+	// Check are present usages matching allowed usages
+	if len(csr.Spec.Usages) != 3 {
+		return "", fmt.Errorf("there are no exactly three usages defined")
 	}
+	for _, usage := range csr.Spec.Usages {
+		if !isUsageInUsageList(usage, allowedUsages) {
+			return "", fmt.Errorf("usage %v is not in the list of allowed usages (%v)", usage, allowedUsages)
+		}
+	}
+
+	return nodeName, nil
+}
+
+// validateX509CSR validates the certificate request by comparing CN with username,
+// and organization with groups.
+func (r *reconciler) validateX509CSR(csr *certificatesv1beta1.CertificateSigningRequest, certReq *x509.CertificateRequest, machine v1alpha1.Machine) error {
 	// Validate Subject CommonName
-	if csrRequest.Subject.CommonName != csr.Spec.Username {
-		return fmt.Errorf("commonName '%s' is different then CSR username '%s'", csrRequest.Subject.CommonName, csr.Spec.Username)
+	if certReq.Subject.CommonName != csr.Spec.Username {
+		return fmt.Errorf("commonName '%s' is different then CSR username '%s'", certReq.Subject.CommonName, csr.Spec.Username)
 	}
+
 	// Validate Subject Organization
-	if len(csrRequest.Subject.Organization) != 1 {
-		return fmt.Errorf("error validating subject organizations")
+	if len(certReq.Subject.Organization) != 1 {
+		return fmt.Errorf("expected only one organization but got %d instead", len(certReq.Subject.Organization))
+	}
+	if certReq.Subject.Organization[0] != nodeGroup {
+		return fmt.Errorf("organization '%s' doesn't match node group '%s'", certReq.Subject.Organization[0], nodeGroup)
+	}
+
+	machineAddressSet := sets.NewString(machine.Status.NodeRef.Name)
+	for _, addr := range machine.Status.Addresses {
+		machineAddressSet.Insert(addr.Address)
 	}
-	if csrRequest.Subject.Organization[0] != "system:nodes" {
-		return fmt.Errorf("expected 'system:nodes' in organization, but got '%s'", csrRequest.Subject.Organization[0])
+
+	// Validate SAN DNS names
+	for _, dns := range certReq.DNSNames {
+		if len(dns) == 0 {
```

Comment:
Thinking better I'm not sure about the utility of this, but it's a minor thing. We can address this later if it worth it.